buildattr64/buildattr64.rst

```diff
@@ -0,0 +1,904 @@
+the meaning of build attributes for relocatable object files only.  The
+presence of ``.ARM.attributes`` sections in a loadable-unit is
+permitted, but the interpretation of the contents of the section is
+outside the scope of this document. The use of ``.note.gnu.property``
+sections for loadable-units is recommended for SysVr4 platforms,
+see (SYSVABI64_) for details.
+
+.. raw:: pdf
+
+   PageBreak
+
+Formal syntax of an ELF Attributes Section
+==========================================
+
+An ELF Attributes section uses a processor specific section with details
+
+.. table:: Attributes Section
+
+    +---------------------+----------------------------+-------+-----------+------+------------+
+    | Name                | Type                       | Flags | Alignment | Info | Entry Size |
+    +=====================+============================+=======+===========+======+============+
+    | ``.ARM.attributes`` | ``SHT_AARCH64_ATTRIBUTES`` | 0x0   | 1         | 0    | 0          |
+    +---------------------+----------------------------+-------+-----------+------+------------+
+
+A consumer may not assume the natural alignment of a data type such as ``uint32``.
+
+The syntactic structure of an attributes section is::
+
+   <format-version: <version-letter>>
+   [ <uint32: subsection-length> NTBS: vendor-name
+     <bytes: vendor-data>
+   ]*
+
+*Format-version* describes the format of the following data. It is a
+single byte. Only version 'A' (0x41) is defined. This field exists to
+allow future changes in format.
+
+*Subsection-length* is a 4-byte integer in the byte order of the ELF
+file. It encodes the length of the subsection, including the length
+field itself, the vendor name string and its terminating NUL byte, and
+the following vendor data. It gives the offset from the start of this
+subsection to the start of the next one.
+
+*Vendor-name* is a NUL-terminated byte string (NTBS) like a C-language
+literal string.
+
+No requirements are placed on the format of private vendor data. The
+format of the public attributes subsections (which will have a
+pseudo-vendor name with a prefix of “aeabi”) are described below.
+
+Attributes that record data about the compatibility of this
+relocatable object file with other relocatable object files must be
+recorded in a public "aeabi" prefixed subsection.
+
+Attributes meaningful only to the producer must be recorded in a
+private vendor subsection. These must not affect compatibility between
+relocatable object files. A producer that does not recognize the
+private vendor may skip over the private vendor subsection.
+
+Formally, there are no constraints on the order or number of vendor
+subsections. A consumer can collect the public ("aeabi" prefixed)
+attributes in a single pass over the section, then all of its private
+data in a second pass.
+
+Public aeabi prefixed subsections
+=================================
+
+About public tags
```

Comment:
Here is a suggestion for this paragraph that reflect my understanding of public subsections:
```suggestion
Public subsections

Subsections are denoted with the tag: `.aeabi_subsection <name>` 
Public subsections are subsections with <name> starting with a prefix of "aeabi".
Public subsections have a header that indicates whether the subsection is mandatory or not, and what is the type of the value for all the subsection's attributes.  

A consumer must be able to parse the header of all the mandatory public subsections. A consumer may skip an optional subsection.
```

Several points:
- The header does not indicates only optionality, but also the attribute's values type.
- The part about the subsection that do not need to be read is confusing. The essence is, if I am not mistaken: If subsection is mandatory, it has to be read, not being able to read it is an error. If it is optional, it can not be read, from any reason (e.g. not being able to recognize the full vendor-name or any other reason, what is special about the vendor-name?)

Reply:
Thanks for the suggestion, I'll have a think about what I can incorporate.

I'm thinking that I could move the assembly support into the main specification as an appendix.